### ISSUE TITLE
I2I: AMP-Iframely component

### ISSUE THREAD

**Issue by iparamonau** [ampproject/amphtml#26226]
Please refer to PR #26151. This I2I describes the component in the PR.

## Summary

AMP support of the widgets provided and hosted by Iframely service (https://iframely.com).

## Motivation

Iframely is SaaS that aggregates widgets and embed codes from third party publishers. We support over 2000 rich media providers, and generate card previews for other URLs. 

Our users include top media companies on the Internet. We generate the embed codes for rich media in their articles and make sure the codes work in the technical environments and formats that customers need. This should include AMP format. Otherwise, our joint users need to spend significant effort for each provider they would like to support.

## Design

The component's design is similar to any other AMP component that integrates 3rd party media: component creates and handles the state of the internal iFrame that encapsulates the media.

Please see PR #26151 for suggested component that addresses the issues:

 - iFrame is bootstrapped from Iframely CDN via a unique content ID. 
 - Alternatively, a combination of Iframely API key and URL can identify the rich media.
 - The component listens and handles sizing and decoration events from the iFrame.
 - The components adds placeholder image when needed.
 - The component doesn't require 3p iFrame. The Iframely SRC format is part of the service and won't change.

## Additional context

At the moment, Iframely provides `amp-iframe` code for the lack of better alternative. It disables some of our features. We have had multiple support requests from our joint users to have a better integration. The examples of known bugs include inability to expand a post since the resizable amp-iframe won't necessarily resize in viewport, and also inability to decorate iFrame/summary card with the borders.

Besides, AMP documentation states that existing AMP components should be preferred over amp-iframe. The amp-iframe component should be considered a fallback.

/cc @ampproject/wg-approvers @alanorozco @nainar

**Comment by alanorozco** [comment]
Thanks for filing this issue.

Some more thoughts coming later, but could you please specify what this component does, that [`amp-iframe` doesn't?](https://go.amp.dev/c/amp-iframe)

**Comment by alanorozco** [comment]
Thanks for filing this issue. More details needed.

> Iframely is SaaS that aggregates widgets and embed codes from third party publishers.

What kind of widgets? Apart from URL construction and resizing, is there any sort of integration with AMP documents specific to the use cases of the widgets?

> [amp-iframe] disables some of our features. 

Could you specify which, or are these mentioned here already?

> The examples of known bugs include inability to expand a post since the resizable amp-iframe won't necessarily resize in viewport

Which scenarios? AMP denies resize requests unless the component is not in viewport to prevent content shifts. This should be true across all components, I'm noticing that #26151 invokes `changeHeight` which is not allowed to be used directly.

> inability to decorate iFrame/summary card with the borders

Could you specify what the summary card is?

Is this styling that's applied from events? If so, that would break some of our layout constraints.

---

I believe for this to justify a component, it should gracefully degrade when sizing requests are denied, we can't make an exception for this and shift content. An overflow button/expand on click might be a good choice. 

In case of border, I'm not sure we should be on board with changing styling dynamically, it may look jarring. What's the use case?

Specifically:

- `border-radius` and `box-shadow` don't change sizing (and therefore do not break layout), so I'm only concerned about the visual effect of these.

- `border` can change sizing and break layout, besides its visual effect.

---

Please also mention sandbox and Content Security Policy differences with `amp-iframe`.

**Comment by kristoferbaxter** [comment]
Following up on @alanorozco's comment, I'd like to see this go to design review.

This appears from my quick glance to violate the 2nd principle listed here (https://amp.dev/about/how-amp-works/) "Size all resources statically".

**Comment by iparamonau** [comment]
Thanks @alanorozco. The suggested component doesn't need to integrate with any other AMP components.  

One of the main Iframely features is ability to find embed code with only canonical URL and API key. So, it's in no way a duplicate of `amp-iframe`. 

We'll need to review what cloud features are missing at the moment and add them in v0.2 as needed. For now, sizing and decorating of the iFrames is causing the most of support requests - so we figured we'd resolve it first (or have something to link our users too for explanation of limitations).

From the implementation standpoint, we could, of course, use 3p iframe and unfurl the content using our standard platform's embed.js script. Many other components use `.changeHeight` without any problem that way. That would resolve both styling of an iframe and changing the height when the content loaded.  But we've figured 3p iframe is to be avoided, and we've implemented everything directly using iFrame. Let us know if 3p is the way to go. We figured you guys try and avoid it if it's feasible.


BTW, our foremost sizing issue that requires to `.changeHeight` is when we need to reduce the height, not expand it. We try and estimate the height for responsive media, and sometimes it's a bit more than needed, depending on the viewport.

**Comment by alanorozco** [comment]
> One of the main Iframely features is ability to find embed code with only canonical URL and API key. So, it's in no way a duplicate of amp-iframe.

This is ergonomics, not functionality, e.g:

```js
    if (this.id) {
      const id = encodeURIComponent(this.id);
      src = `${this.base}${id}/thumbnail?amp=1`;
    } else {
      const url = encodeURIComponent(this.url);
      src = `${this.base}api/thumbnail?url=${url}&key=${this.key}&amp=1`;
    }
```

These can be constructed manually with `amp-iframe`. This is not necessarily a big deal but it's important to consider since we would like to prevent maintaining one-offs if unnecessary.

> From the implementation standpoint, we could, of course, use 3p iframe and unfurl the content using our standard platform's embed.js script. Many other components use .changeHeight without any problem that way. 

This is not related to them being 3p, but because of the constraints of the service provided and how it resizes. We can discuss further for the specifics of this component and come up with a strategy to resize somewhat safely.

> BTW, our foremost sizing issue that requires to `.changeHeight` is when we need to reduce the height, not expand it. 

This is still a content shift. Consider a situation when the component has been scrolled past the viewport and then gets a resize response: in this case the document reading position is shifted upwards and lost.

Aside from height changes, I'm still worried about borders being added dynamically and affecting size. This can be worked around with an inner component element that's sized so the border's width is not added to the outer component's dimensions, **but** the problem of this potentially being visually jarring still stands.

Let's bring this to [design review.](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md)

**Comment by alanorozco** [comment]
(Making a note of the iframe `allow` policy in the code for the open PR:)

```
encrypted-media *;
accelerometer *;
gyroscope *;
picture-in-picture *;
camera *;
microphone *;
autoplay *;
```

We'll have to discuss each of these and their use cases during design review to prevent abuse. This policy is a contract of trust so we need to clarify the potential problems and how the Iframely service handles them in order to allow them.

**Comment by iparamonau** [comment]
@alanorozco sounds good. Will dig the review process doc in a bit and see how to navigate it.

I see we might need a comment for the review meeting. So here it is for the reference.

Frankly, I am having hard time understanding the design issue. If 3p vs regular iFrame is not a problem, than our design is not conceptually any different from the one of [this component](https://amp.dev/documentation/components/amp-embedly-card/), which is a similar service. Our approach with direct iFrame is even more responsible & safe to my taste, as we excluded our platform's main js file that unfurls widgets and instead just copied the needed functions directly into component. But I'm quite OK to match the design exactly to other components that use 3p if that's simpler and more reliable.

In my understanding, AMP already has a whitelist of  3rd party services that are allowed to bypass resizing restrictions of amp-iframe.  Either other services should be allowed to bypass it too (what's the criteria and was it backwardly applied to existing components?), or nobody should. Or better yet, give it as a config option to the end-users.  Again, we receive multiple bug/support reports to the fact that widgets don't resize, so it's a real issue for some people.

In either case, AMP would be able to enforce it in `.changeHeight` method or similar.

**Comment by alanorozco** [comment]
@iparamonau Before we go into further discussion, are you familiar with the `attemptChangeHeight` API? Does it fulfill your use case?

**Comment by iparamonau** [comment]
@alanorozco I remember we did try `attemptChangeHeight`, but can't tell why we dropped it. Will ask the team to deploy the component with attemptChangeHeight tomorrow and let you know how it goes.

**Comment by iparamonau** [comment]
@alanorozco so `attemptChangeHeight` has the same bugs that we are trying to fix, therefore we switched to `changeHeight` that many other components use. But what we missed is an option to catch errors and write callbacks with `attempt...`, which is great. Also, I see that more 3rd party plugins are not using `attemptChangeHeight` then the ones using `changeHeight` directly. 

We'll be more than happy to rely on AMP checks in component's v0.1 and use `attemptChangeHeight`, if that's the only component's design issue remaining. We'll close the PR and submit a new one when it's ready per this discussion. Please confirm this plan or let me know if we rather need to go into review process. 

For v0.2, we would still be interested to learn about the processes of getting whitelisted for `changeHeight`. Let me know if there's one already or if I need to submit a separate ticket for that process. Alternatively, I kind of feel that if AMP switches all existing components to `attemptChangeHeight`, all the issues inside the articles will become obvious and it will give better view on how to improve the function. But, again, that's a topic for a separate ticket.

**Comment by alanorozco** [comment]
> We'll be more than happy to rely on AMP checks in component's v0.1 and use attemptChangeHeight, if that's the only component's design issue remaining. We'll close the PR and submit a new one when it's ready per this discussion. Please confirm this plan or let me know if we rather need to go into review process.

`attemptChangeHeight` doesn't require exceptions to be made, so the pushback comes from employing `changeHeight` directly. So that change unblocks that problem, that sounds good!

_However_, there's two other blocking issues I've mentioned which we haven't addressed. **These need to be taken to design review.**

### 1. Async defined styling props

Those that only affect visually (`box-shadow` and `border-radius`) don't need to block this PR. `border` worries me just because it affects sizing. This can be worked around by inserting the frame in a container *inside* the component's box that has `box-sizing` properties that won't cause the `<amp-iframely>` box to resize. _Please confirm that this is OK to do._

### 2. [`allow` policy](#issuecomment-571670944)

For `autoplay`, do the frames autoplay video with sound?

- If there's some way to guarantee they're always muted when autoplaying, we can go ahead. This doesn't have to be a technical safeguard but we need a way to audit that this is the way the Iframely service behaves.

- If they're always click-to-play, then we can just remove this policy. 

However, how browsers handle `allow` policies in general can change in the future in a way that can affect UX, so you have to be okay with any of these potentially being removed in the future. We can set communication channels for this, we don't want to move things around unexpectedly.

Additionally, embedding iframes have concerns that require security review. Adding @ampproject/wg-security-privacy and @molnarg. 

> we would still be interested to learn about the processes of getting allowed for `changeHeight`.

Simply put, design review. There's details we need to understand about how the service behaves and its performance in order to allow it. I understand the overall issue that the current behavior is unexpected for clients, but keep in mind that we're trying to enforce end-user guarantees here.

I'll start making technical comments on the PR, but I believe the easiest way to address the two remaining concerns is design review.

**Comment by alanorozco** [comment]
**3rd issue: dynamic domain**, which I just noticed.

_This is very important to discuss during design review._

```js
    const domain = this.element.getAttribute('data-domain') || DOMAIN;
    this.base = `https://${domain
      .replace(/^(https?:)?\/\//g, '')
      .replace(/\/.*$/i, '')}/`;
```

Domain is consumed from `data-domain` attribute. There's no allowlist, all are allowed. 

Consequently, `amp-iframely` can be used generically for any page regardless of origin. It allows anything to get around any `amp-iframe` guarantees, including those that `attemptChangeHeight()` tries to ensure when denying resizes.

It also creates an API for any page to have Iframely specific features which we haven't agreed upon benefiting the AMP ecosystem overall, like async styling.

It also gets around the security constraints of `amp-iframe`. For example, **this allows any video coming from any page to autoplay with sound**.

If the domain set cannot be restricted, or kept static, **this component cannot be allowed to do anything that `amp-iframe` doesn't**, including styling, non-static sizing vs. viewport gurantees, resizing and/or overriding the `allow` policy.

**Comment by alanorozco** [comment]
(Noting also that the previous issue is **blocking**.)

**Comment by alanorozco** [comment]
(Noting also that this may be where some of the confusion is coming from other widget components using 3p nested frames.)

Those other components are under the same guarantees as `amp-iframe`, they don't override the `allow` policy and the additional layer creates other security guarantees. The reason other services which use 3p frames are allowed to `changeHeight` directly is related to how the first-level frame chooses to interact with the outer document, regardless of what the 2nd-level frame includes.

By embedding a URL from any domain as the first-level frame, we're creating the issues mentioned in the comment above.

**Comment by iparamonau** [comment]
This is great and constructive, thanks @alanorozco. 

- Good catch regarding the CDN domain. We'll switch to a whitelist of the biggest distributions we need. 
- Autoplay value isn't really required in v0.1 - we make sure everything has click-to-play. We'll remove it for now.
- Border styling - we'll try and test what works for various cards design templates we have. In worst case, I think we'll just be able to hardcode a specific value.

Also, it sounds like if we switched to 3p iframe, we'd be able to claim `changeHeight` easier. Because of this, we'll start reimplementing component to 3p right away in v 0.1. I sense it would also raise less red security flags too. 

The plan is this: we'll switch to 3p and still use `attemptChangeHeight` in v0.1 and then switch to `changeHeight` after design review of a separate intent-to-implement in v0.2 so that this issue doesn't block the component's release. 

_Please let me know if that doesn't sound right and we should come up with `changeHeight` in 3p right away_.

We'll be closing current PR and will open another one when updates are ready and tested.

**Comment by alanorozco** [comment]
@iparamonau

> Border styling - we'll try and test what works for various cards design templates we have. In worst case, I think we'll just be able to hardcode a specific value.

Unfortunately hardcoding the border won't work. Defining a border that's applied on runtime (directly to the `<amp-iframely>` box) can still a) break runtime sizing calculations and b) shift content on the client (there aren't guarantees about the time during the page load that the extension will execute).

I'm confused about why the intermediate container proposal requires design changes?

To be clear, I'm suggesting that the intermediate container to which the border is applied fills the `<amp-iframely>` box. This causes the border to be applied "inwards" rather than outwards, but should still look roughly the same.

Alternatively, you can overlay an empty, fullbleed box with `pointer-events: none` to draw a border that will be applied outwards but won't affect the component element's sizing (with the obvious risk of overlap with adjacent elements).

**Comment by alanorozco** [comment]
@iparamonau Please also make sure to take this to design review. It's good to have these few issues settled now but they need to be exposed to the larger set of contributors regardless.

This one in particular needs to be greenlit by @ampproject/wg-security-privacy since it involves iframes.

**Comment by iparamonau** [comment]
The  PR #26151 has been updated. All the comments and implementation concerns should be incorporated and addressed.

**Comment by alanorozco** [comment]
@iparamonau Have you scheduled a design review session?

**Comment by iparamonau** [comment]
@alanorozco not yet. I'll have to read through the process doc first and try to figure it out.

**Comment by molnarg** [comment]
A couple of things to add from security PoV:
- It must not render in preview mode, just like amp-iframe. An idea: would it be possible to use amp-iframe under the hood instead of <iframe>, so that it can enforce all (current and future) requirement?
- In some modes (amp4email etc.), it'll need to be disabled, but these modes have their own whitelist of components, so it' fine

**Comment by iparamonau** [comment]
@molnarg thanks for the input. "Preview mode" isn't documented anywhere, so we're a little unsure. But we researched and assume you mean the AMP container preview mode for Ads Tag Manager? 

It looks this is handled by `amp-iframe` via this declaration, that only allows `AMP` html format:

`html_format: AMP  # Iframes not allowed in ads.`

If that's the case - suggested [pull request](https://github.com/ampproject/amphtml/pull/26151) for `amp-iframely` already includes the same declaration. Our component should only work in main mode.

**Comment by molnarg** [comment]
Sorry I meant disabling prerender like amp-video does here: https://github.com/ampproject/amphtml/blob/b80c1f2cf9be4189ca5eda479179251b3ccce61e/extensions/amp-video/0.1/amp-video.js#L90

IIRC amp-iframe has a custom integration with prerender mode instead of having just the flag.

**Comment by iparamonau** [comment]
@molnarg For that matter, `prerenderAllowed()` from `baseElement` returns `false`, and we do not override that method. So disabling of prerender should already be covered, it appears?

**Comment by molnarg** [comment]
Looks OK then, but please verify that it really works like that.

**Comment by iparamonau** [comment]
The PR is now merged (thanks!) Shall we close this I2I as well @alanorozco ?
